File: backend/app/core/security.py
"""
Security module — validates Supabase JWT tokens.

The frontend handles auth via Supabase JS. The backend receives the
Supabase access_token as a Bearer header and validates it by calling
Supabase's /auth/v1/user endpoint.
"""

from fastapi import Depends, HTTPException, status
from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer

from .config import settings

bearer_scheme = HTTPBearer()

# Cache to avoid hitting Supabase on every request (simple in-memory)
_user_cache: dict[str, dict] = {}


async def get_current_user(
    credentials: HTTPAuthorizationCredentials = Depends(bearer_scheme),
) -> str:
    """
    Dependency that validates a Supabase access token and returns the user ID.

    Calls Supabase /auth/v1/user to validate the token and get user info.
    """
    token = credentials.credentials

    # Check cache first (tokens are valid for ~1 hour)
    if token in _user_cache:
        return _user_cache[token]

    import httpx

    try:
        async with httpx.AsyncClient() as client:
            res = await client.get(
                f"{settings.supabase_url}/auth/v1/user",
                headers={
                    "Authorization": f"Bearer {token}",
                    "apikey": settings.supabase_anon_key,
                },
                timeout=5.0,
            )

        if res.status_code != 200:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid or expired token",
            )

        user_data = res.json()
        user_id = user_data.get("id", "")
        if not user_id:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid token",
            )

        # Cache this token → user_id mapping
        _user_cache[token] = user_id

        # Keep cache from growing forever
        if len(_user_cache) > 1000:
            _user_cache.clear()

        return user_id

    except HTTPException:
        raise
    except Exception:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Could not validate token",
        )

File: src/server-core/security.ts
import { Request, Response, NextFunction } from "express";
import { config } from "./config";

interface CachedToken {
  userId: string;
  expiresAt: number;
}

// In-memory token cache — 5 min TTL so revoked tokens expire promptly
const userCache = new Map<string, CachedToken>();
const CACHE_TTL_MS = 5 * 60 * 1000;

export interface AuthInfo {
  userId: string;
  token: string;
}

declare global {
  namespace Express {
    interface Request {
      auth?: AuthInfo;
    }
  }
}

export async function authMiddleware(
  req: Request,
  res: Response,
  next: NextFunction
): Promise<void> {
  const authHeader = req.headers.authorization;
  if (!authHeader?.startsWith("Bearer ")) {
    res.status(401).json({ detail: "Missing authorization header" });
    return;
  }

  const token = authHeader.slice(7);

  // Check cache
  const cached = userCache.get(token);
  if (cached && cached.expiresAt > Date.now()) {
    req.auth = { userId: cached.userId, token };
    next();
    return;
  }

  // Stale or missing — remove and re-validate
  userCache.delete(token);

  try {
    const response = await fetch(`${config.supabaseUrl}/auth/v1/user`, {
      headers: {
        Authorization: `Bearer ${token}`,
        apikey: config.supabaseAnonKey,
      },
    });

    if (response.status !== 200) {
      res.status(401).json({ detail: "Invalid or expired token" });
      return;
    }

    const userData = (await response.json()) as { id?: string };
    const userId = userData.id;
    if (!userId) {
      res.status(401).json({ detail: "Invalid token" });
      return;
    }

    userCache.set(token, { userId, expiresAt: Date.now() + CACHE_TTL_MS });
    // Evict oldest entries if cache is getting large
    if (userCache.size > 1000) {
      const now = Date.now();
      for (const [k, v] of userCache) {
        if (v.expiresAt < now) userCache.delete(k);
      }
    }

    req.auth = { userId, token };
    next();
  } catch {
    res.status(401).json({ detail: "Could not validate token" });
  }
}
